File: cli/robo_cli/config/pyproject.py
import tomllib
from functools import lru_cache
from typing import NotRequired, TypedDict

from .validator import validate_schema

TOOL_ROBO = TypedDict(
    "TOOL_ROBO",
    {
        "name": str,
        "description": NotRequired[str],
        "python": NotRequired[str],
        "dependencies": NotRequired[dict[str, str]],
        "dev-dependencies": NotRequired[dict[str, str]],
    },
)


# TODO: Should an absolute path come from some global context/config?
@lru_cache
def load(path="pyproject.toml"):
    with open(path, "rb") as file:
        try:
            content = tomllib.load(file)
            section = content["tool"]["robo"]
        except KeyError as err:
            raise ValueError("Missing tool.robo section in pyproject.toml") from err

        try:
            if validate_schema(section, TOOL_ROBO):
                return section
            raise RuntimeError  # Unreachable code
        except (KeyError, ValueError) as err:
            raise ValueError(f"Invalid configuration in pyproject.toml: {err}") from err

File: cli_go/config/pyproject/pyproject.go
package pyproject

import (
	toml "github.com/pelletier/go-toml/v2"
	"os"
)

type Root struct {
	Tool *Tool `toml:"tool"`
}

type Tool struct {
	Robo *Robo `toml:"robo"`
}

type Robo struct {
	Name            string            `toml:"name"`
	Description     string            `toml:"description"`
	Python          string            `toml:"python"`
	Output          string            `toml:"output"`
	Dependencies    map[string]string `toml:"dependencies"`
	DevDependencies map[string]string `toml:"dev-dependencies"`
}

func LoadPath(path string) (*Robo, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var root Root
	err = toml.Unmarshal(data, &root)
	if err != nil {
		return nil, err
	}

	return root.Tool.Robo, nil
}
